### PR TITLE
Fixes #26735 - web component mounter for react

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Mounts a React component from our `componentRegistry` to the page using web components. Enables mounting by:
```erb
<react-component
  name="ComponentInRegistry"
  data-props="<%= {
    someProp: 'val',
    anotherProp: 'val2'
  }.to_json %>"
></react-component>
```
Packaging: theforeman/foreman-js#26 ~theforeman/foreman-packaging#3711~

**Edit**: the #6734 has been simplified and squashed here and is no longer blocker.

For general discussion please follow [blog](https://community.theforeman.org/t/react-components-mounted-by-web-components/13767).

### CODE REVIEW

**Comment by ohadlevy** [comment]
can we now remove the workaround in turbolinks to unmount components?

**Comment by ohadlevy** [comment]
also, I would expect to add depreciation warnings for the current way of mounting? and changing core to use the new approach ?

**Comment by ezr-ondrej** [comment]
> can we now remove the workaround in turbolinks to unmount components?

Yes as long as all the components switch to the new approach.

> also, I would expect to add depreciation warnings for the current way of mounting? and changing core to use the new approach ?

I was planning to have followup PR, to discuss the approach first and keep it small, but if you think we should roll with the approach, I can do it here :snowboarder:

**Comment by ohadlevy** [comment]
On Sun, May 5, 2019, 16:18 Ondřej Ezr <notifications@github.com> wrote:

> can we now remove the workaround in turbolinks to unmount components?
>
> Yes as long as all the components switch to the new approach.
>
> also, I would expect to add depreciation warnings for the current way of
> mounting? and changing core to use the new approach ?
>
> I was planning to have followup PR, to discuss the approach first and keep
> it small, but if you think we should roll with the approach, I can do it
> here 🏂
>

I would like to add at least few so tests will pick it up, but I really
don't see value in having both approaches at the same time...?

**Comment by ezr-ondrej** [comment]
> I really don't see value in having both approaches at the same time...?

Well there is the thing I do not understand web components properly, so the idea was to iterate it for having it tested first :man_shrugging:

> I would like to add at least few so tests will pick it up

What would you think about adding some obvious but less exposed switches here and a PR to switch all components in core as a followup?

**Comment by ohadlevy** [comment]
On Sun, May 5, 2019, 16:35 Ondřej Ezr <notifications@github.com> wrote:

> I really don't see value in having both approaches at the same time...?
>
> Well there is the thing I do not understand web components properly, so
> the idea was to iterate it for having it tested first 🤷‍♂
>
> I would like to add at least few so tests will pick it up
>
> What would you think about adding some obvious but less exposed switches
> here and a PR to switch all components in core as a followup?
>
Ack, maybe something in the experimental section?

**Comment by ezr-ondrej** [comment]
> Ack, maybe something in the experimental section?

As the experimental section is empty now, I have chosen something bit more exposed.

1. **Power status on host list** - as it is isolated feature, without critical impact, but we will get report soon if it shall not work properly on some browsers/devices
2. **React form input** - only used in `generate_at` in report templates, but as it is pretty new, I believe users are going to try that out.

Makes sense? I left it in separate commit, so there is no problem to change the selection :)

File: app/helpers/tags/react_input.rb
Comment (sharvit) [inline app/helpers/tags/react_input.rb]: `to_json` is not necessary now

File: app/views/hosts/_list.html.erb
Comment (sharvit) [inline app/views/hosts/_list.html.erb]: `to_json` is not necessary now

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @ezr-ondrej :+1:

**Comment by ezr-ondrej** [comment]
@sharvit improved the react helper and component it self a lot - kudos :+1:
and decided to add dates to the party, with a "we should be confident at first shot" tone :)

**Comment by ezr-ondrej** [comment]
There appears to be problem with the tests though. PhantomJS doesn't support this feature and there would be too much of effort to fix it just in testing env. I believe we are blocked by #5770 to go :green_heart:

**Comment by ohadlevy** [comment]
> There appears to be problem with the tests though. PhantomJS doesn't support this feature and there would be too much of effort to fix it just in testing env. I believe we are blocked by #5770 to go

on a very very quick look, I would say the polyfill doesn't work as expected, this is the reason I wanted to see tests running :-)

I would assume that either the phantomjs wont work with this polyfill, which means we have to change our browser compatibility? looking at the errors it looks like you got some actual JS errors (probably for non supported syntax the polyfil didnt cover)

**Comment by ezr-ondrej** [comment]
Ok, I believe that #6748 will make us go :green_heart:, but I wondered if @ohadlevy got a point and it seems, that the webcomponents are working in Edge and IE 11. It seems to stop support `onchange` (and probably all the others `on<event>`) attribute though. What would be another challenge to get rid of those. I just wonder why it is, because I haven't found anything about it while Googling it :man_shrugging:

**Comment by ezr-ondrej** [comment]
I've rebased this and pin pointed the issue, it is just an `onchange` event for select2, which is used only in host form, and it can be easily fixed by using unobtrusive way, so I think, this is actually not even blocked by IE support.

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @ezr-ondrej, great progress !
few thoughts:
1. how the `onchange` event can be fixed?
2. how about adding a unit test that snaps a web component
3. Is there an integration test that checks if a component is mounted correctly?
4. could you update the storybook's documentation in the mounting section?
5. does it affect styling (by adding html element it might break css rules - it's a common issue with web components) ?

**Comment by ezr-ondrej** [comment]
1. #6945 is addressing the issue and we should be good to go if that is merged.
2. and
3. will try to do something about tests
4. yeah I updated the docs accordingly, thanks
5. well if we wrote the styles badly, it can affect them, but it really shouldn't have I belive. It could be so if we would use shadowDOM, than obviously we would need to change CSS, but we are not using it, so in most cases we should be alright, one more element in the path should not matter in most cases I believe :)

**Comment by xprazak2** [comment]
I kicked off Travis again, failures were caused by crashed coverals.  Hound sniffed out one violation though.

**Comment by ezr-ondrej** [comment]
> I kicked off Travis again, failures were caused by crashed coverals. Hound sniffed out one violation though.

I can see it, but it is not a violation, will look for disabling it. It is a name of the custom component tag, so that's intended for him not to know it :)

**Comment by ezr-ondrej** [comment]
I've turned off the rule completely :thinking:

**Comment by xprazak2** [comment]
Any additional comments?

**Review by amirfefer — DISMISSED** [review]
LGTM! thanks @ezr-ondrej  this is very useful!
can you just rebase to see if tests aren't failing?

**Comment by sharvit** [comment]
`@theforeman/vendor` contain the webcomponent polyfills now. Need to update to version `1.4.0` to make it available.
See https://github.com/theforeman/foreman-js/blob/master/CHANGELOG.md#140-2019-08-14

PR in foreman to update the vendor: https://github.com/theforeman/foreman/pull/6988
Packaging PR: https://github.com/theforeman/foreman-packaging/pull/4024

**Comment by sharvit** [comment]
> `@theforeman/vendor` contain the webcomponent polyfills now. Need to update to version `1.4.0` to make it available.
> See https://github.com/theforeman/foreman-js/blob/master/CHANGELOG.md#140-2019-08-14
> 
> PR in foreman to update the vendor: #6988
> Packaging PR: [theforeman/foreman-packaging#4024](https://github.com/theforeman/foreman-packaging/pull/4024)

Both of them are merged so you can rebase

**Comment by tbrisker** [comment]
What's the status here?

**Comment by amirfefer** [comment]
@tbrisker - I think this is ready to be merged, got an ack from me and @sharvit, rebased and  https://github.com/theforeman/foreman-js/pull/26 is merged.

File: app/helpers/reactjs_helper.rb
Comment (tbrisker) [inline app/helpers/reactjs_helper.rb]: it would be better to use kwarg here for `flatten_data` instead of a generic opts object. tbh, i'd even prefer getting rid of that option at least here and deciding on an opinionated way of passing data to components - if the component want some data wrapped in a `data` prop, the caller should wrap the data in that way, which it would need to know about anyways to decide if flatten_data should be true or false. 
Even if we still need it on the react side, we don't have to support it here as well - iirc this was just some workaround because we were inconsistent in the past regarding passing both data and props to components in different ways.

File: app/helpers/reactjs_helper.rb
Comment (tbrisker) [inline app/helpers/reactjs_helper.rb]:
```suggestion
  #          valid value types: Hash, json-string, nil
```

File: app/helpers/reactjs_helper.rb
Comment (tbrisker) [inline app/helpers/reactjs_helper.rb]: why default data to empty array if that is not a valid value type? or if it is, the comment is out of sync with the code

File: app/views/hosts/_list.html.erb
Comment (tbrisker) [inline app/views/hosts/_list.html.erb]: if we drop the `flatten_data` option and default it to true, this would become:
```suggestion
            <%= react_component('PowerStatus', data: {id: host.id, url: power_api_host_path(host)}) %>
```

**Review by tbrisker — CHANGES_REQUESTED** [review]
Tested, most places work well, but the form input layout gets broken with this change:
image
Also, added some comments regarding the `flatten_data` option - tl;dr - i think we should drop it in this implementation at least on the ruby side.

File: app/helpers/reactjs_helper.rb
Comment (tbrisker) [reply app/helpers/reactjs_helper.rb]: maybe even rename the parameter to props so it's clear this will be the component's props.

**Comment by mmoll** [comment]
@ezr-ondrej please rebase as GH-6945 got merged.

**Review by amirfefer — DISMISSED** [review]
Can you fix the linting issue?

**Comment by tbrisker** [comment]
what's the status here @ezr-ondrej ?

**Comment by ezr-ondrej** [comment]
I've got rid of rendering the wrapper tag around the component, so it should not affect styling at all, the onChange handlers shouldn't be affected even on IE as those are extracted and Travis approves (for what it's worth here, but I guess we won't get better coverage). Any more comments about this one, @tbrisker, @amirfefer, @sharvit ?

File: app/helpers/reactjs_helper.rb
Comment (ezr-ondrej) [reply app/helpers/reactjs_helper.rb]: I've done this one and I believe the rest is related. Is it better like this? :)

File: webpack/stories/docs/addingNewComponent.md
Comment (sharvit) [inline webpack/stories/docs/addingNewComponent.md]: I am wondering, why should we allow 2 different ways to achieve the same? Wouldn't it lead to confusion and inconsistency?

Comment (ezr-ondrej) [reply webpack/stories/docs/addingNewComponent.md]: I wanted to show what it's doing under the hood. What about this (not encouraging people to use it, but still showing it)?

File: webpack/stories/docs/addingNewComponent.md
Comment (tbrisker) [inline webpack/stories/docs/addingNewComponent.md]:
```suggestion
Will render the following HTML:
```

File: webpack/stories/docs/addingNewComponent.md
Comment (sharvit) [reply webpack/stories/docs/addingNewComponent.md]: Much better! thanks!

**Review by tbrisker — DISMISSED** [review]
Tested, works well! just a small wording correction in the doc part, other than that good to go.

File: webpack/stories/docs/addingNewComponent.md
Comment (ezr-ondrej) [reply webpack/stories/docs/addingNewComponent.md]: :+1: done

**Review by sharvit — APPROVED** [review]
Thanks @ezr-ondrej, looks good and works as expected :+1:

**Comment by sharvit** [comment]
Merging, thanks @ezr-ondrej and everyone who was involved here.

**Pull request merged**
